modules/replication/module.py

```diff
@@ -65,53 +71,62 @@ def __init__(self, id, resolver, n, f, k):
+            if (not self.rep[self.id][VIEW_CHANGE] and
                     self.resolver.execute(Module.VIEW_ESTABLISHMENT_MODULE,
-                                          Function.ALLOW_SERVICE):
-                view_changed = self.rep[self.id] != self.TEE and \
-                    (self.resolver.execute(
-                        Module.VIEW_ESTABLISHMENT_MODULE,
-                        Function.GET_CURRENT_VIEW, self.id) != self.id)
+                                          Function.ALLOW_SERVICE)):
+                view_changed = (self.rep[self.id] != self.TEE and
+                                (self.resolver.execute(
+                                    Module.VIEW_ESTABLISHMENT_MODULE,
+                                    Function.GET_CURRENT_VIEW, self.id) !=
+                                    self.rep[self.id][PRIM]))
                 self.rep[self.id][VIEW_CHANGE] = view_changed
 
-            self.prim = self.resolver.execute(
+            self.rep[self.id][PRIM] = self.resolver.execute(
                 Module.VIEW_ESTABLISHMENT_MODULE,
                 Function.GET_CURRENT_VIEW, self.id)
 
             # lines 4-6
-            if self.rep[self.id][VIEW_CHANGE] and self.prim == self.id:
+            if (self.rep[self.id][VIEW_CHANGE] and
+                    self.rep[self.id][PRIM] == self.id):
                 # this node is acting as primary
                 processor_ids = []
-                for i, replica_structure in self.rep:
-                    if replica_structure[VIEW_CHANGE] and self.prim == i:
-                        processor_ids.add(i)
-
-                if len(processor_ids) > (3 * self.number_of_byzantine) + 1:
+                for j, replica_structure in enumerate(self.rep):
+                    j_prim = self.resolver.execute(
+                        Module.VIEW_ESTABLISHMENT_MODULE,
+                        Function.GET_CURRENT_VIEW,
+                        j
+                    )
+                    if (replica_structure[VIEW_CHANGE] and
+                            j_prim == self.id):
+                        processor_ids.add(j)
+
+                if len(processor_ids) > (4 * self.number_of_byzantine) + 1:
                     self.renew_reqs(processor_ids)
                     self.find_cons_state(self.com_pref_states(
                         (3 * self.number_of_byzantine) + 1
                     ))  # TODO assign return val from cons_state when impl.
                     self.rep[self.id][VIEW_CHANGE] = False
             # lines 7-8
-            elif self.rep[self.id][VIEW_CHANGE] and \
-                    (self.rep[self.prim][VIEW_CHANGE] is False and
-                     (self.rep[self.id][PRIM] == self.rep[self.prim][PRIM])):
+            elif (self.rep[self.id][VIEW_CHANGE] and
+                    (self.rep[PRIM][VIEW_CHANGE] is False and
+                     (self.rep[self.id][PRIM] ==
+                        self.rep[self.rep[self.id][PRIM]][PRIM]))):
```

Comment:
haha "fy fan", it's correct but "fyfan"

Reply:
:(

![image](https://user-images.githubusercontent.com/6113675/52640451-297aa900-2edf-11e9-9ea1-a93dc8d90bc9.png)